libs/execution/tests/unit/polymorphic_executor.cpp

```diff
@@ -0,0 +1,203 @@
+
+    f.get();    // propagate exceptions
+
+    HPX_TEST_EQ(passed_through, 42);
+    return 42;
+}
+
+void test_then(executor const& exec)
+{
+    hpx::future<void> f = hpx::make_ready_future();
+
+    HPX_TEST(
+        hpx::parallel::execution::then_execute(exec, &test_f, std::move(f), 42)
+            .get() == 42);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+std::atomic<std::size_t> count(0);
+
+int bulk_test(std::size_t value, int passed_through)
+{
+    ++count;
+    HPX_TEST_EQ(passed_through, 42);
+    return passed_through;
+}
+
+void test_bulk_sync(executor const& exec)
+{
+    std::vector<int> v(107);
+    std::iota(v.begin(), v.end(), std::rand());
+
+    using hpx::util::placeholders::_1;
+    using hpx::util::placeholders::_2;
+
+    count = 0;
+    hpx::parallel::execution::bulk_sync_execute(
+        exec, hpx::util::bind(&bulk_test, _1, _2), v, 42);
+    HPX_TEST(count == v.size());
+
+    count = 0;
+    hpx::parallel::execution::bulk_sync_execute(exec, &bulk_test, v, 42);
+    HPX_TEST(count == v.size());
+}
+
+///////////////////////////////////////////////////////////////////////////////
+void test_bulk_async(executor const& exec)
+{
+    std::vector<int> v(107);
+    std::iota(std::begin(v), std::end(v), std::rand());
+
+    using hpx::util::placeholders::_1;
+    using hpx::util::placeholders::_2;
+
+    count = 0;
+    hpx::when_all(hpx::parallel::execution::bulk_async_execute(
+                      exec, hpx::util::bind(&bulk_test, _1, _2), v, 42))
+        .get();
+    HPX_TEST(count == v.size());
+
+    count = 0;
+    hpx::when_all(
+        hpx::parallel::execution::bulk_async_execute(exec, &bulk_test, v, 42))
+        .get();
+    HPX_TEST(count == v.size());
+}
+
+///////////////////////////////////////////////////////////////////////////////
+int bulk_test_f(
+    std::size_t value, hpx::shared_future<void> const& f, int passed_through)
+{
+    HPX_TEST(f.is_ready());    // make sure, future is ready
+
+    f.get();    // propagate exceptions
+
+    ++count;
+    HPX_TEST_EQ(passed_through, 42);
+    return passed_through;
+}
+
+void test_bulk_then(executor const& exec)
+{
+    std::vector<int> v(107);
+    std::iota(std::begin(v), std::end(v), std::rand());
+
+    using hpx::util::placeholders::_1;
+    using hpx::util::placeholders::_2;
+    using hpx::util::placeholders::_3;
+
+    hpx::shared_future<void> f = hpx::make_ready_future();
+
+    count = 0;
+    hpx::parallel::execution::bulk_then_execute(
+        exec, hpx::util::bind(&bulk_test_f, _1, _2, _3), v, f, 42)
+        .get();
+    HPX_TEST(count == v.size());
+
+    count = 0;
+    hpx::parallel::execution::bulk_then_execute(exec, &bulk_test_f, v, f, 42)
+        .get();
+    HPX_TEST(count == v.size());
+}
+
+void static_check_executor()
+{
+    using namespace hpx::traits;
+
+    static_assert(has_sync_execute_member<executor>::value,
+        "!has_sync_execute_member<executor>::value");
```

Comment:
*Really* important:
```suggestion
        "has_sync_execute_member<executor>::value");
```
Feel free to wait for CI and make this change just before merging.

Reply:
How did that happen? :/